components/add.js

```diff
@@ -0,0 +1,306 @@
+import { SaveCancelControl, SavedErrors, sortedJson, useEditor } from "./edit";
+import { useRouter } from "next/router";
+import { empty } from "empty-schema";
+import { PROFILE_COLLECTIONS } from "../lib/profiles";
+import { urlWithoutParams } from "../lib/general";
+
+/**
+ *
+ * Given the schema ID, fetch the corresponding collection
+ * name. Example: schema id /profiles/human_donor.json maps
+ * to "human-donors"
+ */
+export const collectionPath = (schema) => {
+  return PROFILE_COLLECTIONS[schema["$id"]];
+};
+
+/**
+ * Looks for an action in the item with a name in the list of given actions
+ * and returns the profile path of the first one found, or null if none are found.
+ * @param {*} item snovault object
+ * @param {*} actions list of actions to find from which to grab
+ * the profile path
+ * @returns the profile path corresponding to an action or null if there
+ * are no available actions or if the none of the specified actions are present.
+ */
+const actionProfile = (item, actions) => {
+  if ("actions" in item) {
+    const act = item.actions.find((act) => actions.includes(act.name));
+    if (act != null) {
+      return act.profile;
+    }
+  }
+  return null;
+};
+
+/**
+ * This takes a schema object and adds to the required properties
+ * in order for the blank template library to autogenerate stubs
+ * for as many fields as makes sense
+ *
+ * We eliminate properties that have permission requirements, not
+ * submittable, commented with "Do not submit", or arrays/objects.
+ *
+ * @param {*} schema the schema to modify in perpartion for producing
+ * an empty template
+ *
+ * @returns The modified schema with appropriate fields added as
+ * required
+ */
+const convertOptionalIntoRequiredSchema = (schema) => {
+  const topProperties = Object.entries(schema.properties)
+    .filter(([, p]) => {
+      return p.comment ? !p.comment.includes("Do not submit") : true;
+    })
+    .filter(([, p]) => {
+      return !p.permission;
+    })
+    .filter(([, p]) => {
+      return p.notSubmittable ? !p.notSubmittable : true;
+    })
+    .filter(([, p]) => {
+      return !["array", "object"].includes(p.type);
+    })
+    .map(([k]) => {
+      return k;
+    });
+  const newschema = schema;
+  newschema.required = [...topProperties, ...schema.required];
+  return newschema;
+};
+
+/**
+ * Generates a button link to the #!add url for the given collection.
+ * A custom label can be suplied with the `label` prop.
+ */
+export const AddLink = ({
+  collection,
+  label = "Add Instance",
+  type = "primary",
+  size = "sm",
+}) => {
+  const collectPath = urlWithoutParams(collection["@id"]);
+
+  if (canEdit(collection, ["add"])) {
+    return (
+      <Button.Link type={type} size={size} href={`${collectPath}#!add`}>
+        {label}
+      </Button.Link>
+    );
+  }
+};
+
+AddLink.propTypes = {
+  collection: PropTypes.object.isRequired,
+  label: PropTypes.string,
+  type: PropTypes.string,
+  size: PropTypes.string,
+};
```

Comment:
It’s not required, but you can catch more errors is you select the allowed values with [PropTypes.oneOf](https://reactjs.org/docs/typechecking-with-proptypes.html#proptypes)